contributors/design-proposals/mount-options.md

```diff
@@ -0,0 +1,145 @@
+      path: "/var/exports"
+      mountOption: "hard,nolock,nfsvers=3"
+```
+
+Or it can be specified as part of StorageClass definition:
+
+```yaml
+apiVersion: storage.k8s.io/v1beta1
+kind: StorageClass
+metadata:
+  name: slow
+provisioner: kubernetes.io/glusterfs
+parameters:
+  resturl: "http://127.0.0.1:8081"
+  restauthenabled: "true"
+  restuser: "admin"
+  secretNamespace: "default"
+  secretName: "heketi-secret"
+  mountOption: "volume-name=foobar"
+```
+
+## Preventing users from specifying invalid/unsupported mount options
+
+While mount options enable more flexibility in how volumes are mounted, it can result
+in user specifying options that are not supported or are known to be problematic.
+
+There could be several possibly alternatives for this. I am listing few I considered.
+
+
+1. Add support for blacklisting mount options as `kubelet` startup configuration:
+
+To prevent users from using options which are not safe - Kubernetes admin can configure volume options and specify a blacklist of mount point parameters. For example:
+
+```
+~> kubelet --volume-options <volume-config-file>
+```
+
+The volume configuration file may look like:
+
+```
+{
+    "glusterfs": {
+      "mount_option_blacklist":  ["transport"]
+    },
+    "nfs": {
+        "mount_option_blacklist": ["hard", "noacl"]
+    }
+}
+```
+
+To match options that are allowed - we will split the mount options in PV by `','` and then verify if key does not exist in blacklist, which will prevent mounting volumes with these options.
+
+One upside of using these options in kubelet is - it will also work for external provisioners. One major downside is - we will be allowing creation of PV or SC which can be blocked later by a blacklisted option. i.e - A PV or SC with invalid mount option will not be rejected immediately.
+
+2. As suggested by @saad change mount options to key/value pairs such as:
+
+```
+ apiVersion: v1
+  kind: PersistentVolume
+  metadata:
+    name: pv0003
+  spec:
+    capacity:
+      storage: 5Gi
+    accessModes:
+      - ReadWriteOnce
+    persistentVolumeReclaimPolicy: Recycle
+    nfs:
+      path: /tmp
+      server: 172.17.0.2
+      mountOption:
+        hard: 'true'
+        nolock: 'true'
+        nfsvers: 3
+```
+
+The obvious downside is - this will require massive scrutiny of supported mount options by each volume plugin.
+Each time a change is needed, we will have to update API too.
+
+The upside is - it is simpler to implement and we can possibly also validate and document each option separately.
+
+
+3. A third option is - we can still use blob of string to specify mount option but instead of using a admin
+configurable blacklist - each Kubernetes volume plugin defines which mount option keys it supports. A bit like:
+
+```
+func (plugin *volume.Plugin) MountKeys() []string {
+    return ["foo", "bar"]
+}
+```
+
+This still has the problem that - each time a new option has to be supported user will have to
+wait for new release.
```

Comment:
+1 to whitelisting. I propose whitelist would be implemented per volume plugin (set of options allowed will differ between NFS and GCE PD, for example). The whitelist will be checked at mount time, when the mount call happens, unsupported mount options will be dropped (with an warning even generated indicating that the specified options will be ignored since they are not in the whitelist).